packages/nextjs/src/config/webpack.ts

```diff
@@ -108,6 +117,87 @@ export function constructWebpackConfigFunction(
+ * map's paths with the given distribution directory.
+ *
+ * This action is required to view correct stack traces when using a custom
+ * distribution directory. The project root is required to build correct paths of the
+ * files that must be overwritten. This use case arises when you run the build step in
+ * a different directory from the one the project is; for example, in tests.
+ *
+ * Why this approach?
+ * There are two times to define the path: run time and build time. In any case,
+ * Next.js requires the user to set `distDir` in the options to compile the project.
+ * To get this path at run time and the SDK build it, users must also define the
+ * option in the SDK's server's configuration. However, to get the path at build time,
+ * users only need to set it once since the `withSentryConfig` wrapper can read the
+ * user's Next.js configuration. The SDK generates the `RewriteFrames` integration at
+ * initialization, meaning it's impossible to dynamically set the distribution directory
+ * path at build time if the option is only available at run time. To solve this issue,
+ * the SDK at build time rewrites the files where the default path of the integration is,
+ * so it's successfully generated at run time without additional configuration.
+ *
+ * The project root is
+ * @param projectRootDir Root directory of the Next.js project.
+ * @param distDir The distribution directory.
+ */
+function updateRewriteFramesBasepath(projectRootDir: string, distDir: string): void {
+  if (distDir === PROJECT_BASEPATH) return;
+  try {
+    // esm
+    setProjectBasepath(
+      path.join(projectRootDir, 'node_modules', '@sentry', 'nextjs', 'esm', 'index.server.js'),
+      'var ',
+      distDir,
+    );
+    // es5
+    setProjectBasepath(
+      path.join(projectRootDir, 'node_modules', '@sentry', 'nextjs', 'dist', 'index.server.js'),
+      'exports.',
+      distDir,
+    );
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.warn(
+      'Sentry Logger [Warn]: ' +
+        `Could not set custom build directory \`${distDir}\`, required for correct display of source maps. `,
+      'In order to set your own `RewriteFrames`, visit\n' +
+        'https://docs.sentry.io/platforms/javascript/guides/nextjs/configuration/integrations/plugin/#rewriteframes\n',
+      error,
+    );
+  }
+}
+
+/**
+ * Overrides the base path variable with the distDir provided.
+ *
+ * Note that only replaces the first regex match.
+ * Different module formats have different variable prefixes and their output files
+ * are located in different paths.
+ *
+ * There's no error handling.
+ *
+ * @param filePath Path to the file.
+ * @param varPrefix Prefix of the variable declaration.
+ * @param distDir The distribution directory.
+ */
+function setProjectBasepath(filePath: string, varPrefix: string, distDir: string): void {
+  const fileContents = fs.readFileSync(filePath).toString();
```

Comment:
That was my initial approach and afaik we can definitely do it, but since we're running the code before webpack runs, the number of files is small (2 currently, and this is not something that scales), and files themselves are small, I believe the sync approach is faster (even if we block the event loop). That said, if you think the async approach is more convenient, I'm happy to change it. What do you think?

Reply:
I prefer defaulting to async, but I'll leave it up to you to make the final call.